spring-amqp/src/main/java/org/springframework/amqp/support/converter/AbstractJavaTypeMapper.java

```diff
@@ -21,80 +21,101 @@
-            finalIdClassMapping.put(id, clazz);
-            this.classIdMapping.put(clazz, id);
-        }
-        this.idClassMapping = finalIdClassMapping;
-    }
-
-    public Map<String, Class<?>> getIdClassMapping() {
-        return Collections.unmodifiableMap(this.idClassMapping);
-    }
-
-    @Override
+public abstract class AbstractJavaTypeMapper {
+
+	public static final String DEFAULT_CLASSID_FIELD_NAME = "__TypeId__";
+
+	public static final String DEFAULT_CONTENT_CLASSID_FIELD_NAME = "__ContentTypeId__";
+
+	public static final String DEFAULT_KEY_CLASSID_FIELD_NAME = "__KeyTypeId__";
+
+	private final Map<String, Class<?>> idClassMapping = new HashMap<String, Class<?>>();
+
+	private final Map<Class<?>, String> classIdMapping = new HashMap<Class<?>, String>();
+
+	public String getClassIdFieldName() {
+		return DEFAULT_CLASSID_FIELD_NAME;
+	}
+
+	public String getContentClassIdFieldName() {
+		return DEFAULT_CONTENT_CLASSID_FIELD_NAME;
+	}
+
+	public String getKeyClassIdFieldName() {
+		return DEFAULT_KEY_CLASSID_FIELD_NAME;
+	}
+
+	public void setIdClassMapping(Map<String, Class<?>> idClassMapping) {
+		this.idClassMapping.putAll(idClassMapping);
+		createReverseMap();
+	}
+
+	protected void addHeader(MessageProperties properties, String headerName, Class<?> clazz) {
+		if (this.classIdMapping.containsKey(clazz)) {
+			properties.getHeaders().put(headerName, this.classIdMapping.get(clazz));
+		}
+		else {
+			properties.getHeaders().put(headerName, clazz.getName());
+		}
+	}
+
+	protected String retrieveHeader(MessageProperties properties, String headerName) {
+		String classId = retrieveHeaderAsString(properties, headerName);
+		if (classId == null) {
+			throw new MessageConversionException(
+					"failed to convert Message content. Could not resolve " + headerName + " in header");
+		}
+		return classId;
+	}
+
+	protected String retrieveHeaderAsString(MessageProperties properties, String headerName) {
+		Map<String, Object> headers = properties.getHeaders();
+		Object classIdFieldNameValue = headers.get(headerName);
+		String classId = null;
+		if (classIdFieldNameValue != null) {
+			classId = classIdFieldNameValue.toString();
+		}
+		return classId;
+	}
+
+	private void createReverseMap() {
+		this.classIdMapping.clear();
+		for (Map.Entry<String, Class<?>> entry : this.idClassMapping.entrySet()) {
+			String id = entry.getKey();
+			Class<?> clazz = entry.getValue();
+			this.classIdMapping.put(clazz, id);
+		}
+	}
+
+	public Map<String, Class<?>> getIdClassMapping() {
+		return Collections.unmodifiableMap(this.idClassMapping);
+	}
+
+	/**
+	 * @throws Exception an exception.
+	 * @deprecated - no longer necessary.
+	 */
+	@Deprecated
 	public void afterPropertiesSet() throws Exception {
-        validateIdTypeMapping();
-    }
+	}
+
+	protected boolean hasInferredTypeHeader(MessageProperties properties) {
+		boolean hasInferredTypeHeader = properties.getInferredArgumentType() != null;
+		return hasInferredTypeHeader;
```

Comment:
No reason to extract the variable.
The compilator is good here with optimization, but the source code is still this and maybe simplified.